### PR TITLE
Generate Response<bool> return type for head requests when head-as-boolean flag is true

### PR DESCRIPTION
**Pull request by ShivangiReja**
As per [this](https://github.com/Azure/autorest/issues/3603) approach, generating return type as `Response<bool>` for head requests when `head-as-boolean` flag is set to true. For data-plane libraries the default is false and for management libraries the default is true as described [here ](https://github.com/Azure/autorest/issues/3602).

Fixes: https://github.com/Azure/autorest.csharp/issues/912

TODO: Cover all the test scenarios for head requests.

### CODE REVIEW

File: src/AutoRest.CSharp.V3/Generation/Writers/RestClientWriter.cs [hunk 1]
```diff
@@ -517,7 +517,17 @@ private void WriteStatusCodeSwitch(CodeWriter writer, CodeWriterDeclaration mess
                 writer.Line($"var headers = new {headersModelType}({responseVariable});");
             }
 
-            using (writer.Scope($"switch ({responseVariable}.Status)"))
+            string responseStatus;
```
Comment (pakrym) [inline src/AutoRest.CSharp.V3/Generation/Writers/RestClientWriter.cs:520]: We should handle this in the case statement instead.


something like

``` C#
  switch(status)
        {
            case 200:
            case int s when s > 200 && s < 300:
                Console.WriteLine("yay");
                break;
        }
```

File: src/AutoRest.CSharp.V3/Generation/Writers/RestClientWriter.cs [hunk 2]
```diff
@@ -549,6 +566,12 @@ private void WriteStatusCodeSwitch(CodeWriter writer, CodeWriterDeclaration mess
                             writer.Line($"var {valueVariable:D} = {message}.ExtractResponseContent();");
                             value = new Reference(valueVariable.ActualName, responseBody.Type);
                         }
+                        else if (responseBody is ConstantResponseBody _ && returnType != null)
```
Comment (pakrym) [inline src/AutoRest.CSharp.V3/Generation/Writers/RestClientWriter.cs:569]: Don't discard the ConstantResponseBody variable here and avoid the cast on the next line.

File: src/AutoRest.CSharp.V3/Generation/Writers/RestClientWriter.cs [hunk 3]
```diff
@@ -549,6 +566,12 @@ private void WriteStatusCodeSwitch(CodeWriter writer, CodeWriterDeclaration mess
...
+                        {
+                            var body = (ConstantResponseBody)responseBody;
+                            writer.Line($"{returnType} {valueVariable:D} = {body.Value.Value:L};");
```
Comment (pakrym) [inline src/AutoRest.CSharp.V3/Generation/Writers/RestClientWriter.cs:572]: `writer.WriteReferenceOrConstant`

File: src/AutoRest.CSharp.V3/Output/Models/Responses/Response.cs [hunk 4]
```diff
@@ -5,13 +5,46 @@ namespace AutoRest.CSharp.V3.Output.Models.Responses
...
             ResponseBody = responseBody;
             StatusCodes = statusCodes;
         }
 
         public ResponseBody? ResponseBody { get; }
-        public int[] StatusCodes { get; }
+        public StatusCodes[] StatusCodes { get; }
+    }
+
+    internal enum Family
```
Comment (pakrym) [inline src/AutoRest.CSharp.V3/Output/Models/Responses/Response.cs:18]: Us this used?

Comment (pakrym) [reply src/AutoRest.CSharp.V3/Output/Models/Responses/Response.cs:18]: I think we can just keep the numbers.

File: src/AutoRest.CSharp.V3/Output/Models/RestClient.cs [hunk 5]
```diff
@@ -334,14 +335,30 @@ private RestClientMethod BuildMethod(Operation operation, HttpRequest httpReques
 
             foreach (var response in operation.Responses)
             {
+                List<StatusCodes> statusCodes = new List<StatusCodes>();
+                Array.ForEach(response.HttpResponse.IntStatusCodes, statusCode => statusCodes.Add(new StatusCodes(statusCode, null)));
```
Comment (pakrym) [inline src/AutoRest.CSharp.V3/Output/Models/RestClient.cs:339]: use a foreach statement.

File: src/AutoRest.CSharp.V3/Output/Models/Responses/Response.cs [hunk 6]
```diff
@@ -5,13 +5,46 @@ namespace AutoRest.CSharp.V3.Output.Models.Responses
...
+        CLIENT_ERROR,
+
+        // 5xx HTTP status codes.
+        SERVER_ERROR,
+
+        // Other, unrecognized HTTP status codes.
+        OTHER
+    }
+
+    internal struct StatusCodes
```
Comment (pakrym) [inline src/AutoRest.CSharp.V3/Output/Models/Responses/Response.cs:39]: One type per file

**Comment by pakrym** [comment]
Nice!

File: src/AutoRest.CSharp.V3/Generation/Writers/RestClientWriter.cs [hunk 7]
```diff
@@ -542,7 +532,16 @@ private void WriteStatusCodeSwitch(CodeWriter writer, CodeWriterDeclaration mess
                         }
                         else
                         {
-                           writer.Line($"case {statusCode.Family}:");
+                            if (statusCode.Family == 2)
+                            {
+                                writer.Line($"case 200:");
```
Comment (pakrym) [inline src/AutoRest.CSharp.V3/Generation/Writers/RestClientWriter.cs:537]: You don't need this line for the family case. The example had it both ways to show that we can mix the individual codes and ranges.

` writer.Line($"case int s when s >= {statusCode.Family*100:L} && s < {statusCode.Family * 100 + 99:L}:");

File: test/AutoRest.TestServer.Tests/HeadAsBooleanTrue.cs [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+        [Test]
+        public Task HttpSuccess204Head() => Test(async (host, pipeline) =>
+        {
+            var response = await new HttpSuccessClient(ClientDiagnostics, pipeline, host).Head204Async();
+            Assert.AreEqual(204, response.GetRawResponse().Status);
+            Assert.AreEqual(true, response.Value);
+        });
+
+        [Test]
+        public Task HttpSuccess404Head() => Test(async (host, pipeline) =>
```
Comment (pakrym) [inline test/AutoRest.TestServer.Tests/HeadAsBooleanTrue.cs:32]: Do we have an endpoint that returns 3xx?

File: test/AutoRest.TestServer.Tests/HeadAsBooleanTrue.cs [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+using NUnit.Framework;
+
+namespace AutoRest.TestServer.Tests
+{
+    public class HeadAsBooleanTrueTests : TestServerTestBase
+    {
+        public HeadAsBooleanTrueTests(TestServerVersion version) : base(version) { }
+
+        [Test]
+        public Task HttpSuccess200Head() => Test(async (host, pipeline) =>
```
Comment (pakrym) [inline test/AutoRest.TestServer.Tests/HeadAsBooleanTrue.cs:16]: Let's add a test validates the API surface as well (check method return type).

File: test/AutoRest.TestServer.Tests/HeadAsBooleanTrue.cs (hunk 8 above)
Comment (ShivangiReja) [reply test/AutoRest.TestServer.Tests/HeadAsBooleanTrue.cs:32]: We don't have 3xx returns [here](https://github.com/Azure/autorest.testserver/blob/master/swagger/head.json).

Comment (ShivangiReja) [reply test/AutoRest.TestServer.Tests/HeadAsBooleanTrue.cs:32]: But we have routes which return 3xx. Do you want me to update the spec to add use case for 3xx?

Comment (pakrym) [reply test/AutoRest.TestServer.Tests/HeadAsBooleanTrue.cs:32]: It's ok.

File: src/AutoRest.CSharp.V3/Generation/Writers/RestClientWriter.cs (hunk 7 above)
Comment (ShivangiReja) [reply src/AutoRest.CSharp.V3/Generation/Writers/RestClientWriter.cs]: Ahh, yess!

**Review by pakrym — APPROVED** [review]

**Pull request merged**